src/Microsoft.TestPlatform.Client/TestPlatform.cs

```diff
@@ -170,44 +170,23 @@ public TestPlatform()
             this.AddExtensionAssemblies(testSessionCriteria.RunSettings);
 
             var runConfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(testSessionCriteria.RunSettings);
-
-            // Update extension assemblies from source when design mode is false.
-            //
-            // TODO (copoiena): Is it possible for this code to run if we're not in design mode ?
-            // An use case for this would be when running tests with "dotnet test". Usually there's
-            // a build involved then.
             if (!runConfiguration.DesignMode)
             {
-                return;
-            }
-
-            // Initialize loggers.
-            var loggerManager = this.TestEngine.GetLoggerManager(requestData);
-            loggerManager.Initialize(testSessionCriteria.RunSettings);
-
-            var testHostManager = this.testHostProviderManager.GetTestHostManagerByRunConfiguration(testSessionCriteria.RunSettings);
-            ThrowExceptionIfTestHostManagerIsNull(testHostManager, testSessionCriteria.RunSettings);
-
-            testHostManager.Initialize(TestSessionMessageLogger.Instance, testSessionCriteria.RunSettings);
-
-            if (testSessionCriteria.TestHostLauncher != null)
-            {
-                testHostManager.SetCustomLauncher(testSessionCriteria.TestHostLauncher);
+                return false;
             }
 
-            var testSessionManager = this.TestEngine.GetTestSessionManager(requestData, testHostManager, testSessionCriteria);
+            var testSessionManager = this.TestEngine.GetTestSessionManager(requestData, testSessionCriteria);
             if (testSessionManager == null)
             {
                 // The test session manager is null because the combination of runsettings and
                 // sources tells us we should run in-process (i.e. in vstest.console). Because
                 // of this no session will be created because there's no testhost to be launched.
                 // Expecting a subsequent call to execute tests with the same set of parameters.
                 eventsHandler.HandleStartTestSessionComplete(null);
```

Comment:
That's a really good point indeed, thanks for bringing it up.

The event handler is only used to notify TE (who initiated the `StartTestSession` call on `VsTestConsoleWrapper` in the first place) that the operation either succeeded or failed and is currently the only way to do that, the return value doesn't matter at all for this as it can't leave the TP context due to the way it is architected (see `DesignModeClient.cs`). In this case you can look at the event handler as a callback that encapsulates the communication logic that wires `vstest.console` and the `TranslationLayer` together. Passing down the event handler in the call hierarchy enables us to let TE know as soon as possible if something went wrong, because we'd call `HandleStartTestSessionComplete(null)` as soon as a fail condition is encountered. The current implementation is analogous to the way discovery/run operations are implemented.

The return value on the other hand is for TP only. It's easier to extract the result of the `StartTestSession` operation like that than by using the event handler implemented in the fashion described above. In this way the return value is passed up the call hierarchy and is handled where it makes sense by logging an error. The other advantage of having a return value to those operations is that it counts for easier testing, i.e. we can pass a certain combination of parameters to one of the methods and expect it either to succeed of fail and that would be tested by asserting the return value is equal to some other predefined value. If the methods were void it'd be harder to provide comprehensive unit testing.

Reply:
so what happens when TE creates a session that's invalid. will it get `null` back, or will it get an invalid session object that we should discard based on the `HandleStartTestSessionComplete(null)` event?

Reply:
Oh, I think I understand what you meant now. In my previous comment I was talking about the `vstest.console` side-of-things API which in my opinion is properly designed because of the reasons above.

However, on the `TranslationLayer` (i.e. `TE`) side it's not that well designed. We have an events handler as well as a `TestSession` object that's being returned. Since we use the same `ITestSessionEventsHandler` on both sides of the communication (`vstest.console` and `TE`) we must have the `HandleStartTestSessionComplete` method on the interface. However, you can decide to ignore that and instead use the result returned by the `StartTestSession` call on `VsTestConsoleWrapper` (which returns a `TestSession` object) to decide if the operation was successful or not.

Now I'll have to do some changes to the way that object is created because currently, even though the `TestSessionInfo` we get from TP may be `null`, the `TestSession` object won't -- it will instead contain a `null` `TestSessionInfo`. I think a more elegant approach would be to return a `null` `TestSession` if the `TestSessionInfo` we get is `null` too.